config/taxonomy/application_taxonomy_test.go

```diff
@@ -0,0 +1,53 @@
+// Copyright 2020 IBM Corp.
+// SPDX-License-Identifier: Apache-2.0
+
+package taxonomy
+
+import (
+	"testing"
+)
+
+var (
+	m4dAppValsName = "application.values.schema.json"
+
+	intentGood = "{\"intent\":\"Marketing\"}"
+	intentBad  = "{\"intent\":\"whatever\"}"
+
+	roleGood = "{\"role\":\"Data Scientist\"}"
+	roleBad  = "{\"role\":\"whatever\"}"
+
+	// {"app_info":{"intent":"Marketing", "role":"Data Scientist"}}
+	appInfoGood = "{\"app_info\":{\"intent\":\"Marketing\", \"role\":\"Data Scientist\"}}"
+
+	// {"app_info":{"intent":"Marketing", "role":"Data Scientist", "x":"Y"}}
+	appInfoGoodExtraProps = "{\"app_info\":{\"intent\":\"Marketing\", \"role\":\"Data Scientist\", \"x\":\"Y\"}}"
+
+	// {"app_info":{"role":"Data Scientist", "x":"Y"}}
+	appInfoBadNoIntent = "{\"app_info\":{\"role\":\"Data Scientist\", \"x\":\"Y\"}}"
+
+	// {"interface":{"protocol":"m4d-arrow-flight", "data_format":"arrow"}}
+	interfaceGoodFlight = "{\"interface\":{\"protocol\":\"m4d-arrow-flight\", \"data_format\":\"arrow\"}}"
+
+	// {"interface":{"protocol":"m4d-arrow-flight", "data_format":"parquet"}}
+	interfaceBadFlight = "{\"interface\":{\"protocol\":\"m4d-arrow-flight\", \"data_format\":\"parquet\"}}"
+
+	// {"interface":{"protocol":"kafka", "data_format":"json"}}
+	interfaceGoodKafka = "{\"interface\":{\"protocol\":\"kafka\", \"data_format\":\"json\"}}"
+
+	// {"interface":{"protocol":"whatever", "data_format":"avro"}}
+	interfaceBadKafka = "{\"interface\":{\"protocol\":\"whatever\", \"data_format\":\"avro\"}}"
+)
+
+func TestApplicationTaxonomy(t *testing.T) {
```

Comment:
Assuming we have #533 how useful are these tests? 
What are you testing here?

We should have a couple of run scenarios (schemes):
1. M4D with minimal verification, i.e., check only protocol and data_format (with values that are supported today), nothing else. (additional properties True for app_info)
3. M4D with additional check for intent and role accepting specific values as a sample scheme values. Here we should have values as example for intent and role in app_info.
We should proceed in parallel in both scenarios to show that they work.

Reply:
That tool will only validate that the taxonomy defined in a `schema.json` file is in the correct format. The Go tests are to test specific scenarios for values that the user will provide in M4D application and validate that against the taxonomy defined. We are using the [ValidateTaxonomy function](https://github.com/IBM/the-mesh-for-data/blob/9928d662fc48085667f0d340fc2617bdf41d7fbe/config/taxonomy/validate.go#L16) that @simanadler implemented. The Go tests cover both the run scenarios that you listed.

Reply:
Thank you. Would it make more sense to validate in the test with M4Dapplication as an input? the test that were defined in the starting point scheme are somewhat over-simplified.
For example, defining (i) m4dapplication w/o intent/role, (ii) m4dapplication with intent/role
and check if it passes validation?

Reply:
I thought the CRD will already check if the user submits an m4dapplication without intent/role. These Go tests are checking that the intent/role values are one of [these](https://github.com/IBM/the-mesh-for-data/blob/83c8ae31b788b3a78d6db24e9642b3e2b1f678ce/config/taxonomy/policymanager.values.schema.json#L7) defined in the taxonomy.

Reply:
The CRD does not define AppInfo values, it is defined as a map in the CRD,
AppInfo is not used by the manager, so it does not understand its content directly.

My feedback regarding the testing is that it would make more sense to start from a m4dapp data structure and get the values from it, rather than have just one field from the m4dapp structure. That test will evolve to a unit test of the actual code eventually.

Reply:
The intent/role Go tests are similar to the tests [here](https://github.com/IBM/the-mesh-for-data/blob/83c8ae31b788b3a78d6db24e9642b3e2b1f678ce/config/taxonomy/policymanager_taxonomy_test.go#L60). These simply check whether an intent is provided. The `interface` Go tests are checking if the `protocol` and `data_format` are paired up correctly as defined in the taxonomy. 

@ronenkat I see your point. We can have another Go test file which takes an m4dapplication data structure as an input, and validates all the field values with our taxonomy.